### PR TITLE
Add naive first approach to zksnarks as basis for tutorial

### PR DESCRIPTION
**Pull request by iamalwaysuncomfortable**
## Changelog
* Created a naive first approach to zksnarks without encryption of the challenge values/proof of powers used/power shifting to provide an initial basis to show how they work and why encryption and zero knowledge steps are necesssary
* Created snarks suitable for range proofs

## Motivation
It's intended to expose snark libraries to the libraries these are connected to. Providing a explanation of how they operate will help developers & users of this library understand the underyling technology

### CODE REVIEW

File: applied-crypto-examples/src/zksnark.rs [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+        Self {
+            roots,
+            public_roots: Vec::new(),
+        }
+    }
+
+    pub fn set_public_roots(mut self, num_public: usize) -> Self {
+        self.public_roots = self.roots[0..num_public].to_vec();
+        self
+    }
```
Comment (GlitchyBritches) [inline applied-crypto-examples/src/zksnark.rs:74]: This should probably change in the future, but looks good for now.

File: applied-crypto-examples/src/zksnark.rs [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+
+    /// Get degree of polynomial
+    pub fn degree(&self) -> usize {
+        self.roots.len()
+    }
+
+    /// Create public polynomial from private polynomial
+    pub fn get_public_polynomial(&self) -> Polynomial {
+        Polynomial::new(self.public_roots.clone())
+    }
```
Comment (GlitchyBritches) [inline applied-crypto-examples/src/zksnark.rs]: There should be a check here to avoid public polynomials from being generated if no roots are specified

**Review by GlitchyBritches — APPROVED** [review]
A few things to change, but overall good.

File: applied-crypto-examples/src/zksnark.rs (hunk 2 above)
Comment (iamalwaysuncomfortable) [reply applied-crypto-examples/src/zksnark.rs]: Good point. Added.

**Pull request merged**